### PR TITLE
EMBED-1080: Support Embedded Segment Builder

### PR DESCRIPTION
**Pull request by sghosh-cen**
Supporting Embedded Segment Builder!

### CODE REVIEW

File: components/SegmentCreationWizard.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+          if (response.status === "created") {
+            setSegments((segments) => [
+              ...segments,
+              {
+                id: response.details.id,
+              },
+            ])
+            await refetchSegments()
+            // prepares a new link for the next segment creation
+            await resetSegmentManagementLink()
```
Comment (yeldah3) [inline components/SegmentCreationWizard.js:25]: Can we leave this out? Feels weird to "prepare" for another segment session before a user needs one. (I think we do this for syncs too, but that's also funky to me)

File: components/SegmentManagement.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+          ))}
+        {showCreateSegmentWizard ? (
+          <SegmentCreationWizard
+            setSegments={setSegments}
+            refetchSegments={refetchSegments}
+            resetSegmentManagementLink={resetSegmentManagementLink}
+            setShowCreateSegmentWizard={setShowCreateSegmentWizard}
+            connectLink={createLink}
+          />
+        ) : showAddNewSegmentButton ? (
```
Comment (yeldah3) [inline components/SegmentManagement.js:64]: If `showAddNewSegmentButton` is always true can we remove this conditional?

File: hooks/use-segment-management-link.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import { useCallback, useEffect, useState } from "react"
+
+export function useSegmentManagementLink(
+  segmentManagementLinks,
+  refetchSegmentManagementLinks,
+  workspaceAccessToken,
+) {
+  const [now] = useState(() => new Date())
+  const [isLoading, setIsLoading] = useState(false)
+  const [segmentManagementLink, setsegmentManagementLink] = useState(() =>
```
Comment (yeldah3) [inline hooks/use-segment-management-link.js:10]:
```suggestion
  const [segmentManagementLink, setSegmentManagementLink] = useState(() =>
```

File: hooks/use-segment-management-link.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    setsegmentManagementLink(newLink)
+    setIsLoading(false)
+  }, [workspaceAccessToken])
+
+  useEffect(() => {
+    if (segmentManagementLink) return
+    getNewsegmentManagementLink()
+  }, [getNewsegmentManagementLink, segmentManagementLink])
+
+  const resetsegmentManagementLink = async () => {
```
Comment (yeldah3) [inline hooks/use-segment-management-link.js:39]:
```suggestion
  const resetSegmentManagementLink = async () => {
```

File: hooks/use-segment-management-link.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    await refetchSegmentManagementLinks() // This refetches the segment management links. This is necessary as this hook may be unmounted and remounted due to the iframe exiting. This causes the useState to execute again, voiding the setState update.
+    await getNewsegmentManagementLink()
```
Comment (yeldah3) [inline hooks/use-segment-management-link.js:41]:
```suggestion
    await getNewSegmentManagementLink()
```

File: pages/api/list_segments.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  const sources = await getSources(workspaceApiKey)
+  for (const source of sources) {
+    const apiResponse = await fetch(`${censusBaseUrl}/api/v1/sources/${source.id}/filter_segments`, {
+      method: "GET",
+      headers: { ["authorization"]: `Bearer ${workspaceApiKey}` },
+    })
+    await checkStatus(apiResponse, 200)
+    const { data, allowances } = await apiResponse.json()
+    logger.info([data, allowances])
+    allSegments.push(...data)
```
Comment (yeldah3) [inline pages/api/list_segments.js:28]: This also needs to page through all the segments per source right?

File: components/SegmentObject.js [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+            {JSON.stringify(
+              {
+                ["authorization"]: "Bearer <workspaceAccessToken>",
+              },
+              null,
+              2,
+            )}
+          </pre>
+        }
+        link="https://developers.getcensus.com/api-reference/segment-management-links/list-segment-management-links"
```
Comment (yeldah3) [inline components/SegmentObject.js:160]:
```suggestion
        link="https://developers.getcensus.com/api-reference/segment-management-links/create-a-segment-management-link-to-edit-a-segment"
```

File: components/SegmentCreationWizard.js (hunk 1 above)
Comment (sghosh-cen) [reply components/SegmentCreationWizard.js:25]: without this if i create a segment and then try to create another, I get the `Sorry this link has expired`

Comment (yeldah3) [reply components/SegmentCreationWizard.js:25]: Weird. But fine! Fine to leave as is

File: components/Card.js [hunk 8]
```diff
@@ -1,8 +1,9 @@
-export function Card({ className, disabled, children }) {
+export function Card({ className, disabled, children, variant = "default" }) {
```
Comment (henryz2000) [inline components/Card.js:1]: this should probably be exported by default

File: components/EmbeddedFrame.js [hunk 9]
```diff
@@ -13,7 +13,11 @@ export default function EmbeddedFrame({ connectLink, onExit }) {
 
       if (event.data.message === "LOADED") setLoaded(true)
 
-      if (event.data.message === "EXITED_CONNECT_FLOW" || event.data.message === "EXITED_SYNC_FLOW")
+      if (
+        event.data.message === "EXITED_CONNECT_FLOW" ||
+        event.data.message === "EXITED_SYNC_FLOW" ||
+        event.data.message === "EXITED_SEGMENT_FLOW"
+      )
```
Comment (henryz2000) [inline components/EmbeddedFrame.js:20]: we should probably use a `switch` statement or an `else if` clause

File: components/SegmentCreationWizard.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
+import { Card } from "@components/Card"
```
Comment (henryz2000) [inline components/SegmentCreationWizard.js:1]: import default Card

File: components/SegmentCreationWizard.js [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+  resetSegmentManagementLink,
+  setShowCreateSegmentWizard,
+  connectLink,
+}) {
+  return (
+    <Card variant="thin">
+      <EmbeddedFrame
+        connectLink={connectLink}
+        onExit={async (response) => {
+          if (response.status === "created") {
```
Comment (henryz2000) [inline components/SegmentCreationWizard.js:16]: what happens if the response is not "created"?

File: components/SegmentManagement.js [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+          <Button
+            className="flex items-center justify-center rounded-md border border-indigo-500/40 bg-stone-50  px-5 py-8 text-xl
+              shadow-sm"
+            onClick={initiateSegmentWizardFlow}
+          >
+            <a id={`create-segment`}>
```
Comment (henryz2000) [inline components/SegmentManagement.js:70]: NIT: we should only use ticks for computed values (i.e. `"create-segment"`)

File: components/SegmentManagement.js [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+              <i className="fa-solid fa-plus mr-4" />
+              {addNewSegmentText}
+            </a>
+          </Button>
+        ) : null}
+      </div>
+
+      <RequestTooltip
+        anchorSelect={`#create-segment`}
```
Comment (henryz2000) [inline components/SegmentManagement.js:79]: NIT: `"create-segment"`

File: components/SegmentManagement.js [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+                {
+                  redirect_uri: window.location.href,
+                },
+                null,
+                2,
+              )}
+            </pre>
+          )
+        }
+        link="https://developers.getcensus.com/api-reference/segment-management-links/create-a-new-segment-management-link"
```
Comment (henryz2000) [inline components/SegmentManagement.js:108]: let's make this link a constant at the top of the file

File: components/SegmentObject.js [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+}) {
+  const [loading, setLoading] = useState(false)
+  const [disabledOverride, setDisabledOverride] = useState()
+  const [editSegmentWizardLink, setEditSegmentWizardLink] = useState(null)
+  const showEditSegmentWizard = !!editSegmentWizardLink
+
+  const initiateEditSegmentWizard = async () => {
+    try {
+      setLoading(true)
+      const response = await fetch("/api/create_edit_segment_management_link", {
```
Comment (henryz2000) [inline components/SegmentObject.js:25]: links as consts at the top of the file

File: components/SegmentObject.js [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+      }
+    } finally {
+      setLoading(false)
+    }
+  }
+
+  const deleteSegment = async () => {
+    try {
+      setLoading(true)
+      const response = await fetch("/api/delete_segment", {
```
Comment (henryz2000) [inline components/SegmentObject.js:53]: links as consts at the top of the file

File: components/SegmentObject.js [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+              </div>
+            </div>
+          </div>
+        )}
+      </Card>
+
+      <RequestTooltip
+        devMode={devMode}
+        anchorSelect={`#delete-${segment.id}`}
+        url={`${censusBaseUrl}/api/v1/sources/${segment.source_id}/filter_segments/${segment.id}`}
```
Comment (henryz2000) [inline components/SegmentObject.js:129]: links as consts at the top of the file

File: components/SegmentObject.js [hunk 18]
```diff
@@ -0,0 +1,177 @@
...
+            {JSON.stringify(
+              {
+                ["authorization"]: "Bearer <workspaceAccessToken>",
+              },
+              null,
+              2,
+            )}
+          </pre>
+        }
+        link="https://developers.getcensus.com/api-reference/segments/delete-segment"
```
Comment (henryz2000) [inline components/SegmentObject.js:142]: links as consts at the top of the file

File: components/SegmentObject.js [hunk 19]
```diff
@@ -0,0 +1,177 @@
...
+      />
+      <RequestTooltip
+        devMode={devMode}
+        anchorSelect={`#configure-${segment.id}`}
+        url={`${censusBaseUrl}/api/v1/sources/${segment.source_id}/filter_segments/${segment.id}/segment_management_links`}
```
Comment (henryz2000) [inline components/SegmentObject.js:147]: links as consts at the top of the file

File: hooks/use-segment-management-link.js [hunk 20]
```diff
@@ -0,0 +1,45 @@
+import { useCallback, useEffect, useState } from "react"
+
+export function useSegmentManagementLink(
+  segmentManagementLinks,
+  refetchSegmentManagementLinks,
+  workspaceAccessToken,
+) {
+  const [now] = useState(() => new Date())
```
Comment (henryz2000) [inline hooks/use-segment-management-link.js:8]: do we need this to be a state?

File: hooks/use-segment-management-link.js [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+  const [isLoading, setIsLoading] = useState(false)
+  const [segmentManagementLink, setsegmentManagementLink] = useState(() =>
+    segmentManagementLinks.find(
+      (item) => new Date(item.expiration) > now && !item.revoked && !item.expired && !item.segment_id,
+    ),
+  )
+
+  const getNewsegmentManagementLink = useCallback(async () => {
+    setIsLoading(true)
+    const response = await fetch("/api/create_segment_management_link", {
```
Comment (henryz2000) [inline hooks/use-segment-management-link.js:18]: links as consts at the top of the file

File: hooks/use-segment-management-link.js [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+) {
+  const [now] = useState(() => new Date())
+  const [isLoading, setIsLoading] = useState(false)
+  const [segmentManagementLink, setsegmentManagementLink] = useState(() =>
+    segmentManagementLinks.find(
+      (item) => new Date(item.expiration) > now && !item.revoked && !item.expired && !item.segment_id,
+    ),
+  )
+
+  const getNewsegmentManagementLink = useCallback(async () => {
```
Comment (henryz2000) [inline hooks/use-segment-management-link.js:16]: `getNewSegmentManagementLink`

File: hooks/use-segment-management-link.js [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+        ["authorization"]: `Bearer ${workspaceAccessToken}`,
+        ["content-type"]: "application/json",
+      },
+    })
+    if (!response.ok) {
+      setIsLoading(false)
+      throw new Error(response.statusText)
+    }
+    const newLink = await response.json()
+    setsegmentManagementLink(newLink)
```
Comment (henryz2000) [inline hooks/use-segment-management-link.js:30]: `setSegmentManagementLink`

File: hooks/use-segment-management-link.js [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+    setIsLoading(false)
+  }, [workspaceAccessToken])
+
+  useEffect(() => {
+    if (segmentManagementLink) return
+    getNewsegmentManagementLink()
```
Comment (henryz2000) [inline hooks/use-segment-management-link.js:36]: `getNewSegmentManagementLink`

File: hooks/use-segment-management-link.js [hunk 25]
```diff
@@ -0,0 +1,45 @@
...
+  }, [getNewsegmentManagementLink, segmentManagementLink])
+
+  const resetsegmentManagementLink = async () => {
+    await refetchSegmentManagementLinks() // This refetches the segment management links. This is necessary as this hook may be unmounted and remounted due to the iframe exiting. This causes the useState to execute again, voiding the setState update.
+    await getNewsegmentManagementLink()
+  }
+
+  return [segmentManagementLink, resetsegmentManagementLink, isLoading]
```
Comment (henryz2000) [inline hooks/use-segment-management-link.js:44]: `resetSegmentManagementLink`

File: pages/_app.js [hunk 26]
```diff
@@ -141,6 +141,22 @@ function MainApplication({ Component, pageProps, workspaceAccessToken, onLogOut
...
+
+  const {
+    loading: segmentsLoading,
+    error: segmentsError,
+    data: segments,
+    setData: setSegments,
+    refetchInBackground: refetchSegments,
+  } = useBasicFetch(
+    () =>
+      new Request(`/api/list_segments`, {
```
Comment (henryz2000) [inline pages/_app.js:153]: links as consts at the top of the file

File: pages/api/create_edit_segment_management_link.js [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+export default async function handler(req, res) {
+  if (req.method !== "POST") {
+    res.status(405).json({})
+    return
+  }
+
+  const { sourceId, segmentId } = req.body
+  const workspaceApiKey = getWorkspaceAccessToken(req)
+  const apiResponse = await fetch(
+    `${censusBaseUrl}/api/v1/sources/${sourceId}/filter_segments/${segmentId}/segment_management_links`,
```
Comment (henryz2000) [inline pages/api/create_edit_segment_management_link.js:19]: links as consts at the top of the file

File: pages/api/create_segment_management_link.js [hunk 28]
```diff
@@ -0,0 +1,27 @@
...
+const logger = pino({ name: __filename })
+
+export default async function handler(req, res) {
+  if (req.method !== "POST") {
+    res.status(405).json({})
+    return
+  }
+
+  const workspaceApiKey = getWorkspaceAccessToken(req)
+  const apiResponse = await fetch(`${censusBaseUrl}/api/v1/segment_management_links`, {
```
Comment (henryz2000) [inline pages/api/create_segment_management_link.js:17]: links as consts at the top of the file

File: pages/api/delete_segment.js [hunk 29]
```diff
@@ -0,0 +1,24 @@
...
+export default async function handler(req, res) {
+  if (req.method !== "DELETE") {
+    res.status(405).json({})
+    return
+  }
+
+  const { sourceId, segmentId } = req.body
+  const workspaceApiKey = getWorkspaceAccessToken(req)
+  const apiResponse = await fetch(
+    `${censusBaseUrl}/api/v1/sources/${sourceId}/filter_segments/${segmentId}`,
```
Comment (henryz2000) [inline pages/api/delete_segment.js:16]: links as consts at the top of the file

File: pages/api/list_segments.js [hunk 30]
```diff
@@ -0,0 +1,49 @@
...
+    res.status(405).json({})
+    return
+  }
+
+  const allSegments = []
+  const workspaceApiKey = getWorkspaceAccessToken(req)
+
+  const sources = await getSources(workspaceApiKey)
+  for (const source of sources) {
+    const apiResponse = await fetch(`${censusBaseUrl}/api/v1/sources/${source.id}/filter_segments`, {
```
Comment (henryz2000) [inline pages/api/list_segments.js:21]: links as consts at the top of the file

File: pages/api/list_segments.js (hunk 6 above)
Comment (henryz2000) [inline pages/api/list_segments.js:28]: `allSegments` is a const, does this actually push objects to it?

File: pages/api/list_segments.js [hunk 31]
```diff
@@ -0,0 +1,49 @@
...
+  const allData = []
+  while (page) {
+    const apiResponse = await fetch(`${censusBaseUrl}/api/v1/sources`, {
+      method: "GET",
+      headers: { ["authorization"]: `Bearer ${workspaceApiKey}` },
+    })
+    await checkStatus(apiResponse, 200)
+    const { pagination, data } = await apiResponse.json()
+    logger.info([data])
+    allData.push(...data)
```
Comment (henryz2000) [inline pages/api/list_segments.js:45]: `allData` is a const?

File: components/SegmentObject.js [hunk 32]
```diff
@@ -0,0 +1,177 @@
...
+
+  const initiateEditSegmentWizard = async () => {
+    try {
+      setLoading(true)
+      const response = await fetch("/api/create_edit_segment_management_link", {
+        method: "POST",
+        headers: {
+          ["authorization"]: `Bearer ${workspaceAccessToken}`,
+          ["content-type"]: "application/json",
+        },
```
Comment (henryz2000) [inline components/SegmentObject.js:30]: lets make this a variable at the top since we reuse it

File: components/SegmentObject.js [hunk 33]
```diff
@@ -0,0 +1,177 @@
...
+        headers={
+          <pre>
+            {JSON.stringify(
+              {
+                ["authorization"]: "Bearer <workspaceAccessToken>",
+              },
+              null,
+              2,
+            )}
+          </pre>
```
Comment (henryz2000) [inline components/SegmentObject.js:140]: lets make this a variable at the top since we reuse it

File: pages/api/list_segments.js [hunk 34]
```diff
@@ -0,0 +1,49 @@
...
+  }
+
+  const allSegments = []
+  const workspaceApiKey = getWorkspaceAccessToken(req)
+
+  const sources = await getSources(workspaceApiKey)
+  for (const source of sources) {
+    const apiResponse = await fetch(`${censusBaseUrl}/api/v1/sources/${source.id}/filter_segments`, {
+      method: "GET",
+      headers: { ["authorization"]: `Bearer ${workspaceApiKey}` },
```
Comment (henryz2000) [inline pages/api/list_segments.js:23]: lets make this a variable at the top since we reuse it

**Review by henryz2000 — COMMENTED** [review]
let's use a base url and append to it when nested routes are needed.
otherwise looks good!

(nb: sorry for all of the nit comments!)

**Review by yeldah3 — APPROVED** [review]

File: pages/api/list_segments.js (hunk 6 above)
Comment (sghosh-cen) [reply pages/api/list_segments.js:30]: Yes! AllSegments is a const so we cannot assign another array to it but it is not immutable! push still pushes objects

**Pull request merged**